### ISSUE TITLE
Review

### ISSUE THREAD

**Issue by tmcw** [mapbox/geojson-coordinate-properties#1]
@sgillies @lbud @lxbarth @planemad 

Would love a review of this spec.

Some doubts & thoughts here:

Versus the de-facto implementation in toGeoJSON, this nests the property one level further and tries to generalize it beyond heart rates & times.

This does not encode values in coordinates as some have suggested ( https://github.com/geojson/draft-geojson/issues/57 ). I think nested properties are more likely to survive round-trips and are more straightforward to describe than that approach: encoding values in coordinates would require some kind of sidebar property that define what each coordinate position meant, whereas nested properties puts the kind in the key name.

**Comment by lbud** [comment]
I like this. The one thing I wonder about is 

> A Feature object with coordinateProperties defined in its properties MUST contain a geometry object of type LineString.

Since other geometry types use arrays of coordinates, and this proposal can be used for any coordinate properties, why not make it extensible for use in other geometries?

**Comment by tmcw** [comment]
Possibly - to support other geometry types, we'd match their levels of nesting, since polygons, multipolys, multilines, etc, all have >1 level of array nesting.

**Comment by anandthakker** [comment]
:+1: to allowing the other geometry types and matching their level of nesting.  Plenty of good use cases, I think, for coordinate properties on multilines and polygons.

**Comment by sgillies** [comment]
New object items (not new array elements) are JSON's natural extension points. I definitely prefer this to extending coordinate arrays.

I think we need to clarify what the union of coordinates and coordProperties is. Is it a Feature or something else? I'm a tiny bit concerned about people interpreting these things as Features and publishing a coordPropertied LineString feature (for example) instead of a collection of Point features.

**Comment by tmcw** [comment]
> I'm a tiny bit concerned about people interpreting these things as Features and publishing a coordPropertied LineString feature (for example) instead of a collection of Point features.

True; the way I think of this is that the union of coordinates and coordProperties is an **extended coordinate**, not a feature. Meaning a linestring with coordproperties is geometrically a linestring and not equivalent to a collection of point features, since the point features have no necessary order or geometric / ordered relationship with each other.

**Comment by hughbris** [comment]
You asked for thoughts and I've brewed this a few weeks, so its length is a byproduct of that – strap in! At the end is a TL;DR.

Like @bhousel, I've also recently discovered this proposal. That's kind of weird because maybe 6 months ago I thought I did a pretty thorough search for something like it, probably using search terms like "timestamped geojson". Maybe I should find out about this "google" thing everyone bangs on about. From following some links in these issue threads, I know that I got close but probably didn't burrow down far enough into some wider threads to see links to here. I started my search when I noticed that [leaflet-omnivore](https://github.com/mapbox/leaflet-omnivore) and other [GPX converters](https://github.com/mpetazzoni/leaflet-gpx) were discarding GPX timestamps. I realised they were not bugs only after looking closely at the GeoJSON spec.

_I felt that there was a strong enough need (particularly in the IoT world), that I actually proposed holding a SotM 2016 workshop to propose such a standard. Maybe my proposal wasn't accepted because of this work, so I feel slightly klutzy on that account :~/ - thanks Tom! :P_

In any case, happy to see this very workable proposal. I implemented it on the storage side in an [activity tracking project I'm involved in](https://github.com/Fabriko/bikebell), and guess I will need to write something now to consume it using Leaflet. I'm keen to know which other projects have developed consumers or producers for this extension.

As well as thinking this is important, it's kind of urgent. I foresee only wild west scenarios if something isn't accepted as standard very soon – disastrous for developers and end users alike. Put me down to steer away from there.

I'm not crazy about a fragile alignment of properties based on array position, but given the apparent constraints on extending GeoJSON position constructs, and the very real possibility of client breakage flagged by @tmcw, it's the best of a bad bunch of (shoehorning) options I've seen proposed. I didn't originally like the descriptor "coordinateProperties" but am now at peace with it.

Just to divert on extended properties not being available _as supplementary elements_ in GeoJSON's position element array, I'm not entirely sure it's true. I found this [in the standard itself](http://geojson.org/geojson-spec.html#positions):

> _Any number of additional elements are allowed -- interpretation and meaning of additional elements is beyond the scope of this specification._

[Tom backed that up in 2015](http://www.macwright.org/2015/03/23/geojson-second-bite.html#position):

> _Note that GeoJSON doesn’t limit you to longitude, latitude, elevation: positions can have four or more coordinates with undefined meaning, like time, rotation, or anything else expressible as a number._

However, even if strictly allowed, an extended array is an unregulated, undocumented situation and it limits us to numeric array member data. Not mutually understood, expressive enough, or robust. So, unacceptable.

I don't think it helps geojson-coordinate-properties to enumerate property semantics other than `times` – so I would lose `heart` (as soul-crushing as that sounds), and replace it with zero or more labelled arrays of arbitrary properties, similar to the way GeoJSON uses `properties`, except that the one called "times" is reserved. I've noticed others have documented (and [implemented](https://github.com/dret/GeoJSON-X)) interpretive discovery of extension fields (links to ontologies etc) in other threads and comments.

Now and finally(!) we get to my overarching realisation after chewing this over for a few days – from an information model perspective, we're doing it wrong! Of course that's astronaut talk, so I support this urgent shim for practical reasons, and because the stakes of not standardising soon are very high. But we _are_ trying to model two different things.

**I think timestamps don't belong in GeoJSON** – it's for descriptive geospatial features, not events that occurred at them. This is backed up in the [GeoJSON introduction](http://geojson.org/geojson-spec.html#introduction):

> _GeoJSON is a format for encoding a variety of geographic data structures._

The axis of GeoJSON is _geography_. The stable base of a log of (static) sensor readings (for example) is the _temporal dimension_. And a **track** is a ''record of a journey'', its focus is on _time+place_ pairings, though it could be conflated with a generalised location-agnostic log. A track is sequenced with timestamps, but there's no need for geospatial primitives beyond `Point`s. A `LineString` is something you can derive/extract from a journey record.

So in a perfect world, we would have 2 or 3 standard formats:
- GeoJSON for geometries,
- simple timestamped logs for arbitrary data – think weather stations (could include and subsume the next one if geospatial points are included),
- journey records: aligned timestamps, points, and arbitrary metadata (e.g. tagged animal movement, transport activity).

Not even sure it's useful to mention this impossible nirvana, but at least you see how clever I am, or at least the painful depths to which I think. It's not going to happen. Just couldn't avoid noting what IMHO is the elephant in the room!

I think the answer to @sgillies' [valid concern about Point abuse](https://github.com/mapbox/geojson-coordinate-properties/issues/1#issuecomment-111777821) is in the above semantic interpretation. `coordinateProperties` is really `coordinateTemporalProperties` in the way this spec wants them used. `properties` should only relate to the geographic object in an ongoing/stable way, not to an events or readings taken at that point at a time. Thus it's definable semantic abuse to use it for static properties on GeoJSON Points.

**TL;DR**: The need for this is real and pressing. geojson-coordinate-properties pretty much fills the need without harm, however shouldn't enumerate arbitrary properties like `heart`. I can close my eyes and imagine a set of related, more focused information models that would leave GeoJSON to do its intended job, but winter is coming and we need to get behind something for now. A critical mass of supporting/adopting applications will strengthen this as an accepted standard.

**Comment by tmcw** [comment]
@hughbris I'm open to the idea of separate formats, but want to keep this discussion along the principles of the IETF: "rough consensus and running code" - so, please include code examples of non-concrete proposals, as a way of making them clear to everyone else and also sussing out the potential gap between theoretical value and reality.

> coordinateTemporalProperties

I'm not sure I agree with this: time is a recommended property, but calling out the time dimension as the primary dimension makes this implementation unnecessarily narrow; a line with elevation data - potentially _without time_ is just as valid and also somewhat common. Whereas `coordinateProperties` specifies the two fundamental elements of the specification: coordinates, matched with properties.

> I don't think it helps geojson-coordinate-properties to enumerate property semantics other than times – so I would lose heart

I understand that this specificity may not be absolutely required, but I don't follow any reasoning that it's detrimental. I think having a commonly-understood meaning of certain properties would improve interoperability with few negative side-effects.

**Comment by hughbris** [comment]
Thanks @tmcw.

> please include code examples of non-concrete proposals

I'm not suggesting we pursue separate standards except as backburner/aspirational work, so I would have thought use cases would suffice for now. I may be missing nuance in your response.

I take your point about helping readers understand, so I'll think on some concrete examples.

> calling out the time dimension as the primary dimension makes this implementation unnecessarily narrow

Agree to disagree maybe on this one. I really don't think you are describing geospatial objects when you add time – space becomes metadata for your event.

> I understand that this specificity may not be absolutely required, but I don't follow any reasoning that it's detrimental.

It's more that it's not a good look. It seems arbitrary, like you had a very specific use case in mind when creating it (you did), rather than considering the wider use cases. (I don't mean that to sound like you were thoughtless, can't think how to rephrase it :-) ). I guess if I was adding my own property, I might think it's some sort of second class citizen.

I seem to be applying the adage of sticking to what's necessary, which is just the extension mechanism.

> I think having a commonly-understood meaning of certain properties would improve interoperability with few negative side-effects.

I think that's what [GeoJSON-X](https://github.com/dret/GeoJSON-X/tree/master/extensions) is trying to solve. That seems a more flexible solution for semantics, in concept even if you don't like its implementation. What you are proposing is a little like hardcoding it.

**Comment by bhousel** [comment]
> Tom backed that up in 2015:
> 
> > Note that GeoJSON doesn’t limit you to longitude, latitude, elevation: positions can have four or more coordinates with undefined meaning, like time, rotation, or anything else expressible as a number.

Probably worth mentioning here that the [RFC7946 spec](https://tools.ietf.org/html/rfc7946#section-3.1.1) contains this language, which limits you to 3 element coordinates (this published in August of 2016):

> Implementations SHOULD NOT extend positions beyond three elements
>    because the semantics of extra elements are unspecified and
>    ambiguous.  Historically, some implementations have used a fourth
>    element to carry a linear referencing measure (sometimes denoted as
>    "M") or a numerical timestamp, but in most situations a parser will
>    not be able to properly interpret these values.  The interpretation
>    and meaning of additional elements is beyond the scope of this
>    specification, and additional elements MAY be ignored by parsers.

**Comment by tmcw** [comment]
Thanks for the insights here - we've collected good feedback and will meditate on it.

**Comment by hughbris** [comment]
Nooo ..

Have been meaning to check back on this issue recently and especially wanted to encourage input about the special property 'heart' from some of the other regular participants. I've meditated quite a bit on it :) I have further notes from my meditations, let me outline them on a quieter day, assuming I find them :/

What about a "last call", one more week for additional feedback? I appreciate you/we want to move on this. I think that's accepted practice for standards development.

**Comment by hughbris** [comment]
Looks like the fabled notes were on my phone which I spent last Christmas Day rebuilding, realising too late the Notes app wasn't syncing properly :/ Unusual instance of data loss for me.

I'm pretty sure the main points were about unnecessary complexity and structural consistency.

I'm going to leave my objections unelaborated, as I have more pressing issues. Ultimately it's a good practice issue, mostly harmless.

I will add a late suggestion I don't think has been mentioned - the spec should recommend a property at the top level to identify the version of this spec used. That way, you more easily allow yourself changes once this is released and used in the wild. _(sssh, don't tell anyone we're taking inspiration from XML ..)_

Thanks for coming up with this and putting it out there, @tmcw, I'm already using it in anger and expect to much more in the near future. I look forward to some Turf JS convenience methods :)
